### PR TITLE
GH-42130: [GLib] Fix building gir files with MSVC

### PR DESCRIPTION
**Pull request by adamreeve**
### Rationale for this change

This is a blocker for #41885

### What changes are included in this PR?

* Adds `*_AVAILABLE_IN_ALL` macros to generated enum functions
* Defines `*_COMPILATION` variables when compiling C code, not just C++
* Fixes checking for MSVC in version.h to work when compiling C code as well as C++ code
* Sets the PATH variable to allow loading required dlls when running the binary generated by g-ir-scanner

### Are these changes tested?

Yes (build is run in CI)

### Are there any user-facing changes?

No
* GitHub Issue: #42130

### CODE REVIEW

**Comment by kou** [comment]
How about splitting "run Ruby tests (not yet working)" to another issue/PR?

Ruby on Windows uses MSYS2 and it includes GLib via glib2 gem. So it may conflict with GLib from vcpkg.

File: .github/workflows/ruby.yml [hunk 1]
```diff
@@ -364,7 +366,8 @@ jobs:
       - name: Install vcpkg
         shell: bash
         run: |
-          ci/scripts/install_vcpkg.sh "${VCPKG_ROOT}"
+          # Install a more recent vcpkg than what's in .env, as we need recent fixes to the gobject-introspection port
+          ci/scripts/install_vcpkg.sh "${VCPKG_ROOT}" 2024.04.26 "$(pwd)/ci/vcpkg/ports-2024.04.26.patch"
```
Comment (adamreeve) [inline .github/workflows/ruby.yml:370]: Should we just update the vcpkg version in `.env` instead? I wasn't sure what the impact of doing that might be.

**Comment by adamreeve** [comment]
> How about splitting "run Ruby tests (not yet working)" to another issue/PR?
> 
> Ruby on Windows uses MSYS2 and it includes GLib via glib2 gem. So it may conflict with GLib from vcpkg.

Yeah I had thought it might be simple to get the tests running after this change but it looks like that's not the case, so I agree that could be a separate PR.

File: .github/workflows/ruby.yml (hunk 1 above)
Comment (kou) [reply .github/workflows/ruby.yml:370]: Yes.
Could you split this to a separated PR because wheel/jar builds also depend on vcpkg.

Comment (adamreeve) [reply .github/workflows/ruby.yml:370]: I've opened #42167 for this, this has caused some build failures for the Python/Java builds which may take a little time to work through

**Comment by adamreeve** [comment]
I've rebased this after the upgrade to vcpkg on main. The only test failures are the "Could not find Protobuf" errors in macos builds that are also happening on main.

**Review by kou — APPROVED** [review]
+1

File: ci/scripts/c_glib_build.sh [hunk 2]
```diff
@@ -34,15 +34,20 @@ if [ -n "${MSYSTEM:-}" ]; then
...
+if [ -n "${VCPKG_ROOT:-}" -a -n "${VCPKG_TRIPLET:-}" ]; then
   vcpkg_install_root="${build_root}/vcpkg_installed"
   $VCPKG_ROOT/vcpkg install --x-manifest-root=${source_dir} --x-install-root=${vcpkg_install_root}
-  export PKG_CONFIG="${vcpkg_install_root}/x64-windows/tools/pkgconf/pkgconf.exe"
-  meson_pkg_config_path="${vcpkg_install_root}/x64-windows/lib/pkgconfig:${meson_pkg_config_path}"
+  export PKG_CONFIG="${vcpkg_install_root}/${VCPKG_TRIPLET}/tools/pkgconf/pkgconf.exe"
+  meson_pkg_config_path="${vcpkg_install_root}/${VCPKG_TRIPLET}/lib/pkgconfig:${meson_pkg_config_path}"
+  # Configure PATH for libraries required by the gobject-introspection generated binary
+  cpp_vcpkg_install_root="${build_root}/cpp/vcpkg_installed"
+  export PATH="${vcpkg_install_root}/${VCPKG_TRIPLET}/bin:${cpp_vcpkg_install_root}/${VCPKG_TRIPLET}/bin:${cpp_vcpkg_install_root}/${VCPKG_TRIPLET}/debug/bin:${PATH}"
```
[the remark this answers is not in the document]
Comment (adamreeve) [reply ci/scripts/c_glib_build.sh]: Sure, that's done now thanks

**Pull request merged**